src/nexusrpc/handler/_cancellation.py

```diff
@@ -0,0 +1,52 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+
+
+class OperationTaskCancellation(ABC):
+    """
+    Indicates whether a a Nexus task has been cancelled during a sync operation or before an async operation has
+    returned a token.
+
+    Nexus worker implementations are expected to provide an implementation that enables
+    cooperative cancellation for both sync and async operation handlers.
+    """
+
+    @abstractmethod
+    def is_cancelled(self) -> bool:
+        """Return True if the associated task has been cancelled."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def cancellation_details(self) -> Optional[str]:
+        """Provide additional context for the cancellation, if available."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def wait_until_cancelled(self, timeout: Optional[float] = None) -> bool:
+        """Block until cancellation occurs or the optional timeout elapses."""
+        raise NotImplementedError
+
+    @abstractmethod
+    async def wait_until_cancelled_async(self) -> None:
```

Comment:
I think these should be flipped and have `wait_until_cancelled` be the async and add `_sync` to the sync form